events/reporter.go

```diff
@@ -0,0 +1,430 @@
+			log.Info("reported receipt: %v", r)
+		} else {
+			select {
+			case reporter.channelReceipt <- r:
+				log.Info("reported receipt: %v", r)
+			default:
+				log.Info("not reporting receipt as no one is listening: %v", r)
+			}
+		}
+	}
+}
+
+func ReportAccountUpdate(a types.Address) {
+	if reporter != nil {
+		if reporter.blocking {
+			reporter.channelAccount <- a
+			log.Info("reported account update: %v", a)
+		} else {
+			select {
+			case reporter.channelAccount <- a:
+				log.Info("reported account update: %v", a)
+			default:
+				log.Info("not reporting account update as no one is listening: %v", a)
+			}
+		}
+	}
+}
+
+// GetNewTxChannel returns a channel of new transactions
+func GetNewTxChannel() chan TransactionWithValidity {
+	if reporter != nil {
+		return reporter.channelTransaction
+	}
+	return nil
+}
+
+// GetActivationsChannel returns a channel of activations
+func GetActivationsChannel() chan *types.ActivationTx {
+	if reporter != nil {
+		return reporter.channelActivation
+	}
+	return nil
+}
+
+// GetLayerChannel returns a channel of all layer data
+func GetLayerChannel() chan NewLayer {
+	if reporter != nil {
+		return reporter.channelLayer
+	}
+	return nil
+}
+
+// GetErrorChannel returns a channel for node errors
+func GetErrorChannel() chan NodeError {
+	if reporter != nil {
+		return reporter.channelError
+	}
+	return nil
+}
+
+// GetStatusChannel returns a channel for node status messages
+func GetStatusChannel() chan struct{} {
+	if reporter != nil {
+		return reporter.channelStatus
+	}
+	return nil
+}
+
+// GetAccountChannel returns a channel for account data updates
+func GetAccountChannel() chan types.Address {
+	if reporter != nil {
+		return reporter.channelAccount
+	}
+	return nil
+}
+
+func GetRewardChannel() chan Reward {
+	if reporter != nil {
+		return reporter.channelReward
+	}
+	return nil
+}
+
+func GetReceiptChannel() chan TxReceipt {
+	if reporter != nil {
+		return reporter.channelReceipt
+	}
+	return nil
+}
+
+// InitializeEventReporter initializes the event reporting interface
+func InitializeEventReporter(url string) error {
+	// By default use zero-buffer channels and non-blocking.
+	return InitializeEventReporterWithOptions(url, 0, false)
+}
+
+// InitializeEventReporterWithOptions initializes the event reporting interface with
+// a nonzero channel buffer. This is useful for testing, where we want reporting to
+// block.
+func InitializeEventReporterWithOptions(url string, bufsize int, blocking bool) error {
+	if reporter != nil {
+		return errors.New("reporter is already initialized, call CloseEventReporter before reinitializing")
+	}
+	reporter = newEventReporter(bufsize, blocking)
+	if url != "" {
+		InitializeEventPubsub(url)
+	}
+	return nil
+}
+
+// SubscribeToLayers is used to track and report automatically every time a
+// new layer is reached.
+func SubscribeToLayers(newLayerCh timesync.LayerTimer) {
+	if reporter != nil {
+		// This will block, so run in a goroutine
```

Comment:
who calls this function? can it leak go routines?

Reply:
Good question. `newLayerCh` comes from `clock.Subscribe()` which uses a non-blocking send. This service is set up the same way as all of the other services which subscribe to the layer ticker. I think this code is okay. LMK if I'm missing anything.